src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1018 @@
+
+  static const std::string KEY_NAME("cluster_lock");
+  //static constexpr unsigned MAX_LOCK_DURATION_SEC = 3600; // 1 hour
+  static constexpr unsigned EPOCH_MAX_LOCK_DURATION_SEC = 30;
+  static constexpr unsigned MAX_LOCK_DURATION_SEC = 60;
+  static const ceph::bufferlist null_bl;
+
+  const uint64_t URGENT_MSG_ID_VAL = 0x00FFFFFFFFFFFFFFULL;
+
+  //---------------------------------------------------------------------------
+  static int set_cluster_id(char buff[], unsigned buff_size, uint64_t id)
+  {
+    char hex_string[sizeof(id)*2+1];
+    memset(hex_string, 0, sizeof(hex_string));
+    unsigned n = snprintf(hex_string, sizeof(hex_string), "%016lx", id);
+    if (n != 16) {
+      std::cerr << "n=" << n << std::endl;
+    }
+    ceph_assert(buff_size <= n);
+    memset(buff, 0, buff_size);
+    unsigned copy_size = std::min(buff_size, n) - 1;
+    memcpy(buff, hex_string, copy_size);
+    return copy_size;
+  }
+
+  //---------------------------------------------------------------------------
+  static std::string get_urgent_msg_cluster_id()
+  {
+    char buff[15];
+    int n = set_cluster_id(buff, sizeof(buff), URGENT_MSG_ID_VAL);
+    std::string cluster_id(buff, n);
+    return cluster_id;
+  }
+
+  //---------------------------------------------------------------------------
+  void cluster::assign_cluster_id()
+  {
+    // choose a 14 Bytes Random ID (so strings won't use dynamic allocation)
+    // Seed with a real random value, if available
+    std::random_device r;
+    // Choose a random mean between 1 ULLONG_MAX
+    std::default_random_engine e1(r());
+    std::uniform_int_distribution<uint64_t> uniform_dist(1, std::numeric_limits<uint64_t>::max());
+    uint64_t rand_high = uniform_dist(e1);
+    while (rand_high == URGENT_MSG_ID_VAL) {
+      rand_high = uniform_dist(e1);
+    }
+    char buff[15];
+    int n = set_cluster_id(buff, sizeof(buff), rand_high);
+    d_cluster_id.insert(0, buff, n);
+
+    ldpp_dout(dpp, 1) << __func__ << "::" << d_cluster_id << "::" << dendl;
+  }
+
+  //---------------------------------------------------------------------------
+  int init_dedup_pool_ioctx(RGWRados                 *rados,
+			    const DoutPrefixProvider *dpp,
+			    librados::IoCtx          *p_ioctx)
+  {
+    rgw_pool dedup_pool(DEDUP_POOL_NAME);
+    std::string pool_name(DEDUP_POOL_NAME);
+    // using Replica-1 for the intermediate data
+    // since it can be regenerated in case of a failure
+    std::string replica_count(std::to_string(1));
+    librados::bufferlist inbl;
+    std::string output;
+    std::string command = R"(
+    {
+      "prefix": "osd pool create",
+      "pool": ")" + pool_name +
+      R"(",
+      "pool_type": "replicated",
+      "size": )" + replica_count +
+      R"(
+    })";
+    int ret = rados->get_rados_handle()->mon_command(command, inbl, nullptr, &output);
+    if (output.length()) {
+      if (output != "pool 'rgw_dedup_pool' already exists") {
```

Comment:
can we check that the existing pool is replica 1?
if not, will it be considered an error?